### DISCUSSION TITLE
Client reuse in documentation

### DISCUSSION THREAD

**Discussion by emanuelef** [aws/aws-sdk-go-v2#1813] [Q&A] [answered]
Hi,
I'm looking for some documentation about the Client reuse.
On the v1 in the Docs I can read:
"Sessions should be cached when possible, because creating a new Session will load all configuration values from the environment, and config files each time the Session is created."

What's the approach with the Client in v2 ? I can see that if I cache it then subsequent calls to ExecuteStatement take less time but I'm wondering if it's possible that the Client can get stale and how to detect that.

Let's say that I have a µservice interacting with DynamoDB, what if I cache the client when starting up with:
`dynamoClient_instance = dynamodb.NewFromConfig(cfg)`

and then using:
`dynamoClient_instance.ExecuteStatement`

Would it be possible that dynamoClient_instance becomes stale ? In case is there a way to check it before running ExecuteStatement ?

Thanks for any help.

**Comment by RanVaknin** [comment]
Hi @emanuelef ,

I'm not sure I 100% understand what you are asking. However If I understand you correctly:

The `newFromDefaultConifg()` call, reads the environment variables or shared config file once - creates a copy and initializes a config once per call. So those will be "fresh" unless you call again `newFromDefaultConifg()` and create a new client again.

The aws Config object is not dynamically produced - so it will remained "cached" in that sense until you construct a new config for your client.

hope that's helpful.
Thanks,
Ran~

**Comment by emanuelef** [comment]
Thanks @RanVaknin,
What I want to be sure is that if I keep the same instance of `dynamoClient_instance` that won't get stale, meaning that `dynamoClient_instance.ExecuteStatement` will always retry and no need to reinstantiate a new `newFromDefaultConifg()` in case of issues connecting.
From your answer I think I could instantiate a new client at the startup of my µservice and never re-istantiate again (of course unless there are config changes but in that case the µservice will be redeployed and restarted anyway).
Having a single instance I notice is saving a lot of time whenever there's a call to the service using `ExecuteStatement` (no need to call `newFromDefaultConifg()` for every request).
What I was worried about is the fact that maybe for any reason that client could get stale and needs a re-instantiation in case.

**Comment by RanVaknin** [comment, marked as answer]
>What I want to be sure is that if I keep the same instance of dynamoClient_instance that won't get stale, meaning that dynamoClient_instance.ExecuteStatement will always retry and no need to reinstantiate a new newFromDefaultConifg() in case of issues connecting.

Lets break this down: 

>dynamoClient_instance.ExecuteStatement will always retry

The client will only retry if two things happen. An error is returned from the service, and that error is retryable.

>no need to reinstantiate a new newFromDefaultConifg() in case of issues connecting.

That is correct. You just want to construct a client once. 
The thing that could get stale is - depending on how you obtain credentials, they might get stale and you'll need to retrieve new fresh credentials. If you are using a shared credentials file to store your static credentials or environment variables this won't be an issue and the SDK client will be "fresh"

Just construct the client once and you will be fine.

**Reply by emanuelef** [reply]
Thanks, should that be mentioned in the official documentation ?
